nixos/modules/services/computing/slurm/slurm.nix

```diff
@@ -3,57 +3,58 @@
 with lib;
 
 let
+  settingsFormat = let
+    listSep = ",";
+    baseTypes = with types; [ bool int float str ];
+    entryType = with types; oneOf ([ (listOf (oneOf baseTypes)) ] ++ baseTypes);
+    settingsType = with types; oneOf [ (attrsOf (attrsOf entryType)) entryType ];
+
+    entryToString = key: val:
+        if isAttrs val then concatStringsSep "\n" (mapAttrsToList (subKey: subVal: "${key}=${subKey} ${valueToString subVal}") val)
+        else "${key}=${valueToString val}";
+
+    valueToString = val:
+        if isList val then concatStringsSep listSep (map (x: valueToString x) val)
+        else if isBool val then (if val then "YES" else "NO")
+        else if isAttrs val then concatStringsSep " " (mapAttrsToList (k: v: "${k}=${valueToString v}") val)
+        else toString val;
+
+    in {
+      type = types.attrsOf settingsType // {
+        description = "Format of slurm config files.";
+      };
+
+      generate = name: value:
+        pkgs.writeTextDir name (( concatStringsSep "\n" (
+          mapAttrsToList (key: val: entryToString key val ) value )) + "\n");
+   };
 
   cfg = config.services.slurm;
   opt = options.services.slurm;
   # configuration file can be generated by http://slurm.schedmd.com/configurator.html
 
   defaultUser = "slurm";
 
-  configFile = pkgs.writeTextDir "slurm.conf"
-    ''
-      ClusterName=${cfg.clusterName}
-      StateSaveLocation=${cfg.stateSaveLocation}
-      SlurmUser=${cfg.user}
-      ${optionalString (cfg.controlMachine != null) "controlMachine=${cfg.controlMachine}"}
-      ${optionalString (cfg.controlAddr != null) "controlAddr=${cfg.controlAddr}"}
-      ${toString (map (x: "NodeName=${x}\n") cfg.nodeName)}
-      ${toString (map (x: "PartitionName=${x}\n") cfg.partitionName)}
-      PlugStackConfig=${plugStackConfig}/plugstack.conf
-      ProctrackType=${cfg.procTrackType}
-      ${cfg.extraConfig}
-    '';
+  configFile = settingsFormat.generate "slurm.conf" cfg.settings;
```

Comment:
To ease transition and offer escape hatches (as I know of one complicated NixOS module where it would have been great to have this), I wish this could be an actual option overrideable by the user who could just bypass `settings` and write the configuration file itself and the settings would be the defaults.

Plus, an assertion could be done to prevent settings & config file being written.

Once this is done, I think it would be fine to merge it. :)

Reply:
I appreciate the change, but it was not what I suggested — adding a `extraConfig` is not the same as adding an escape hatch that replaces the whole `configFile` and is an acceptable alternative to RFC42 rather than stringly-typed `extraConfig` IMHO (because the user can plug his generator).

Reply:
Alright. I put the `extraConfig` back in. Its content now gets appended. You can now override `settings` completely by just setting it to an empty set.

Reply:
Thank you a lot for your patience!

Reply:
OK, I think I see what you mean now. Maybe as it is right now, it is a good solution? Using an alternative  `configFile` adds more complexity as we need to define `StateSaveLocation` and `SlurmSpoolDir` separately, since they are needed by the service.

Reply:
I think it's more or less fine, the only problem with the current solution is that it's not clear we can sidestep completely all the settings, but I think it's fine for now.